### PR TITLE
Update for SFML 2.5.0

### PR DESCRIPTION
**Pull request by spacechase0**
This PR updates CSFML for a few changes in SFML 2.5.0. It builds with SFML's CMake changes, and adds `sf::Clipboard`, `sf::Cursor`, and `sf::VertexBuffer`. It also adds the `sf::Music` loop point support, as well as the `sf::Ftp::upload()` append parameter.

I was unsure how to implement `sfClipboard_get(Unicode)String` since I couldn't find anything that matched 100%, so I mimicked `sfText_get(Unicode)String`, but using global variables in an anonymous namespace.

I was also unsure about converting `sf::VertexBuffer`'s interface, and was a little confused/lazy about it. Let me know what changes should be made.

Code I used for testing the new classes:
```c
#include <SFML/Graphics.h>
#include <SFML/Window.h>

int main(int argc, char* argv[])
{
    sfRenderWindow* window = sfRenderWindow_create( sfVideoMode_getDesktopMode(), "Title", sfDefaultStyle, NULL );

    sfCursor* cursor = sfCursor_createFromSystem( sfCursorHand );

    sfVertexBuffer* vb = sfVertexBuffer_create( 3, sfTriangles, sfVertexBufferStatic );
    sfVertex vertices[3];
    vertices[0].position.x = 0;
    vertices[0].position.y = 0;
    vertices[0].color = sfRed;
    vertices[1].position.x = 0;
    vertices[1].position.y = 32;
    vertices[1].color = sfGreen;
    vertices[2].position.x = 32;
    vertices[2].position.y = 32;
    vertices[2].color = sfBlue;
    sfVertexBuffer_update( vb, vertices, 3, 0 );

    while ( sfRenderWindow_isOpen( window ) )
    {
        sfEvent event;
        while ( sfRenderWindow_pollEvent( window, &event ) )
        {
            if ( event.type == sfEvtClosed )
            {
                sfRenderWindow_close( window );
            }
            else if ( event.type == sfEvtMouseButtonPressed )
            {
                printf( "Clipboard: %s\n", sfClipboard_getString() );
                sfClipboard_setString( "Meow" );
            }
        }

        sfRenderWindow_clear( window, sfWhite );
        sfRenderWindow_drawVertexBuffer( window, vb, NULL );
        sfRenderWindow_display( window );
    }

    sfCursor_destroy( cursor );
    sfRenderWindow_destroy( window );

    return 0;
}
```

### CODE REVIEW

**Comment by svenstaro** [comment]
Wher `SFML/Window/ClipboardStruct.h` supposed to be?

**Comment by ilovezfs** [comment]
@spacechase0 yes I think @svenstaro is right and the ClipboardStruct.h file is missing from your PR

**Comment by spacechase0** [comment]
Oops, that include isn't supposed to exist. Must be a copy/paste artifact. I'll go fix it now.

**Comment by jcowgill** [comment]
One minor thing - I think the `include_directories(${SFML_INCLUDE_DIR})` line from `src/SFML/CMakeLists.txt` can be removed because that variable is not defined by the SFML CMake config scripts.

**Comment by LaurentGomila** [comment]
Yes, the CMake setup has changed in SFML 2.5 and thus CSFML build files should be updated.

**Comment by ilovezfs** [comment]
Any news here? I've been blocking upgrading SFML in Homebrew to 2.5 on this.

**Comment by ilovezfs** [comment]
@spacechase0 I'm seeing
```
bash-3.2$ make -j1
/usr/local/Cellar/cmake/3.11.2/bin/cmake -H/tmp/csfml-20180526-94840-1pp590p -B/tmp/csfml-20180526-94840-1pp590p --check-build-system CMakeFiles/Makefile.cmake 0
/usr/local/Cellar/cmake/3.11.2/bin/cmake -E cmake_progress_start /tmp/csfml-20180526-94840-1pp590p/CMakeFiles /tmp/csfml-20180526-94840-1pp590p/CMakeFiles/progress.marks
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f CMakeFiles/Makefile2 all
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f src/SFML/System/CMakeFiles/csfml-system.dir/build.make src/SFML/System/CMakeFiles/csfml-system.dir/depend
cd /tmp/csfml-20180526-94840-1pp590p && /usr/local/Cellar/cmake/3.11.2/bin/cmake -E cmake_depends "Unix Makefiles" /tmp/csfml-20180526-94840-1pp590p /tmp/csfml-20180526-94840-1pp590p/src/SFML/System /tmp/csfml-20180526-94840-1pp590p /tmp/csfml-20180526-94840-1pp590p/src/SFML/System /tmp/csfml-20180526-94840-1pp590p/src/SFML/System/CMakeFiles/csfml-system.dir/DependInfo.cmake --color=
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f src/SFML/System/CMakeFiles/csfml-system.dir/build.make src/SFML/System/CMakeFiles/csfml-system.dir/build
make[2]: Nothing to be done for `src/SFML/System/CMakeFiles/csfml-system.dir/build'.
[ 10%] Built target csfml-system
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f src/SFML/Window/CMakeFiles/csfml-window.dir/build.make src/SFML/Window/CMakeFiles/csfml-window.dir/depend
cd /tmp/csfml-20180526-94840-1pp590p && /usr/local/Cellar/cmake/3.11.2/bin/cmake -E cmake_depends "Unix Makefiles" /tmp/csfml-20180526-94840-1pp590p /tmp/csfml-20180526-94840-1pp590p/src/SFML/Window /tmp/csfml-20180526-94840-1pp590p /tmp/csfml-20180526-94840-1pp590p/src/SFML/Window /tmp/csfml-20180526-94840-1pp590p/src/SFML/Window/CMakeFiles/csfml-window.dir/DependInfo.cmake --color=
/Applications/Xcode.app/Contents/Developer/usr/bin/make -f src/SFML/Window/CMakeFiles/csfml-window.dir/build.make src/SFML/Window/CMakeFiles/csfml-window.dir/build
[ 12%] Building CXX object src/SFML/Window/CMakeFiles/csfml-window.dir/Clipboard.cpp.o
cd /tmp/csfml-20180526-94840-1pp590p/src/SFML/Window && /usr/local/Homebrew/Library/Homebrew/shims/super/clang++  -DCSFML_EXPORTS -DCSFML_WINDOW_EXPORTS -I/tmp/csfml-20180526-94840-1pp590p/include -I/tmp/csfml-20180526-94840-1pp590p/src -isystem /usr/local/include  -DNDEBUG -fPIC   -o CMakeFiles/csfml-window.dir/Clipboard.cpp.o -c /tmp/csfml-20180526-94840-1pp590p/src/SFML/Window/Clipboard.cpp
/tmp/csfml-20180526-94840-1pp590p/src/SFML/Window/Clipboard.cpp:34:5: error: use of undeclared
      identifier 'sf'
    sf::String ClipboardString;
    ^
/tmp/csfml-20180526-94840-1pp590p/src/SFML/Window/Clipboard.cpp:35:5: error: use of undeclared
      identifier 'std'
    std::string ClipboardStringAnsi;
    ^
/tmp/csfml-20180526-94840-1pp590p/src/SFML/Window/Clipboard.cpp:42:27: error: use of undeclared
      identifier 'sf'
    ClipboardStringAnsi = sf::Clipboard::getString().toAnsiString();
                          ^
/tmp/csfml-20180526-94840-1pp590p/src/SFML/Window/Clipboard.cpp:50:23: error: use of undeclared
      identifier 'sf'
    ClipboardString = sf::Clipboard::getString();
                      ^
/tmp/csfml-20180526-94840-1pp590p/src/SFML/Window/Clipboard.cpp:58:5: error: use of undeclared
      identifier 'sf'
    sf::Clipboard::setString(text);
    ^
/tmp/csfml-20180526-94840-1pp590p/src/SFML/Window/Clipboard.cpp:65:5: error: use of undeclared
      identifier 'sf'
    sf::Clipboard::setString(text);
    ^
6 errors generated.
make[2]: *** [src/SFML/Window/CMakeFiles/csfml-window.dir/Clipboard.cpp.o] Error 1
make[1]: *** [src/SFML/Window/CMakeFiles/csfml-window.dir/all] Error 2
make: *** [all] Error 2
bash-3.2$ 
```

**Comment by texus** [comment]
`#include <SFML/System/Vector2.h>` has to be added on top of `include/SFML/Window/Cursor.h` and `#include <SFML/Window/Clipboard.hpp>` on top of `src/SFML/Window/Clipboard.cpp` to compile the code.

The `include_directories(${SFML_INCLUDE_DIR})` in `src/SFML/CMakeLists.txt` can be removed as already mentioned, but the find_package line above can also be shortened to `find_package(SFML 2 COMPONENTS network graphics audio REQUIRED)`.

**Comment by spacechase0** [comment]
Sorry it took me awhile, I've been a bit busy.

I'm still unhappy with `sfVertexBuffer`, so let me know what I could do better there.

File: include/SFML/Graphics/RenderTexture.h [hunk 1]
```diff
@@ -51,6 +52,18 @@
 ////////////////////////////////////////////////////////////
 CSFML_GRAPHICS_API sfRenderTexture* sfRenderTexture_create(unsigned int width, unsigned int height, sfBool depthBuffer);
```
[the remark this answers is not in the document]
Comment (dabbertorres) [reply include/SFML/Graphics/RenderTexture.h]: I marked stuff as deprecated in the 2.4 update, so marking as \deprecated should be correct.

ie: https://github.com/SFML/CSFML/blob/master/include/SFML/Window/Event.h#L122

File: include/SFML/Graphics/RenderTexture.h [hunk 2]
```diff
@@ -205,6 +218,7 @@ CSFML_GRAPHICS_API void sfRenderTexture_drawCircleShape(sfRenderTexture* renderT
 CSFML_GRAPHICS_API void sfRenderTexture_drawConvexShape(sfRenderTexture* renderTexture, const sfConvexShape* object, const sfRenderStates* states);
 CSFML_GRAPHICS_API void sfRenderTexture_drawRectangleShape(sfRenderTexture* renderTexture, const sfRectangleShape* object, const sfRenderStates* states);
 CSFML_GRAPHICS_API void sfRenderTexture_drawVertexArray(sfRenderTexture* renderTexture, const sfVertexArray* object, const sfRenderStates* states);
+CSFML_GRAPHICS_API void sfRenderTexture_drawVertexBuffer(sfRenderTexture* renderTexture, const sfVertexBuffer* object, const sfRenderStates* states);
```
[the remark this answers is not in the document]
Comment (spacechase0) [reply include/SFML/Graphics/RenderTexture.h]: I wasn't sure about the naming. Are you saying to name it `sfRenderTexture_drawSomething`?

File: include/SFML/Graphics/VertexBuffer.h [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+/// This function should always be called before using
+/// the vertex buffer features. If it returns false, then
+/// any attempt to use sf::VertexBuffer will fail.
+///
+/// \return True if vertex buffers are supported, false otherwise
+///
+////////////////////////////////////////////////////////////
+CSFML_GRAPHICS_API sfBool sfVertexBuffer_isAvailable();
+
+
```
[the remark this answers is not in the document]
Comment (spacechase0) [reply include/SFML/Graphics/VertexBuffer.h]: I got confused with the multiple overloads for the constructor and `update`. I was waiting for feedback on what functions should be added for it.

File: include/SFML/Graphics/RenderTexture.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (eXpl0it3r) [reply include/SFML/Graphics/RenderTexture.h]: No, there are two new functions added in SFML. Maybe we don't need to support both in CSFML, I just wanted to point it out. 🙂

File: include/SFML/Graphics/RenderWindow.h [hunk 4]
```diff
@@ -445,6 +445,7 @@ CSFML_GRAPHICS_API void sfRenderWindow_drawCircleShape(sfRenderWindow* renderWin
 CSFML_GRAPHICS_API void sfRenderWindow_drawConvexShape(sfRenderWindow* renderWindow, const sfConvexShape* object, const sfRenderStates* states);
 CSFML_GRAPHICS_API void sfRenderWindow_drawRectangleShape(sfRenderWindow* renderWindow, const sfRectangleShape* object, const sfRenderStates* states);
 CSFML_GRAPHICS_API void sfRenderWindow_drawVertexArray(sfRenderWindow* renderWindow, const sfVertexArray* object, const sfRenderStates* states);
+CSFML_GRAPHICS_API void sfRenderWindow_drawVertexBuffer(sfRenderWindow* renderWindow, const sfVertexBuffer* object, const sfRenderStates* states);
```
[the remark this answers is not in the document]
Comment (spacechase0) [reply include/SFML/Graphics/RenderWindow.h]: sf::RenderWindow doesn't appear to have this function, just sf::RenderTexture

[the remark this answers is not in the document]
Comment (eXpl0it3r) [reply include/SFML/Graphics/RenderWindow.h]: Woops, I thought it was part of RenderTarget, must have misread

File: include/SFML/Graphics/RenderTexture.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (spacechase0) [reply include/SFML/Graphics/RenderTexture.h]: There's no equivalent to drawing a subset of a vertex array (or this [this](https://www.sfml-dev.org/documentation/2.5.0/classsf_1_1RenderTarget.php#a976bc94057799eb9f8a18ac5fdfd9b73)), so I would lean towards not. I could add both if that would be preferred, I'm just not good at coming up with names. 😛

File: include/SFML/Window/Keyboard.h [hunk 5]
```diff
@@ -140,7 +140,15 @@ typedef enum
     sfKeyF15,          ///< The F15 key
     sfKeyPause,        ///< The Pause key
 
-    sfKeyCount      ///< Keep last -- the total number of keyboard keys
+    sfKeyCount,      ///< Keep last -- the total number of keyboard keys
```
Comment (eXpl0it3r) [inline include/SFML/Window/Keyboard.h]: Minor detail, not fully indented.

**Comment by LaurentGomila** [comment]
Ok, let's merge this :+1:

**Pull request merged**